In [70]:
import pandas as pd
import matplotlib.pyplot as plt

In [71]:
pd.options.mode.chained_assignment = None  # 경고 무시

In [73]:
df = pd.read_excel("2023년.xlsx", sheet_name="가계부 내역")
df

날짜        시간  타입    대분류   소분류               내용      금액   화폐  \
0    2023-12-31  19:21:35  지출     식비    외식        빌라 드 코스테스  -65000  KRW   
1    2023-12-31  19:20:35  이체  내계좌이체   미분류              장보경  100000  KRW   
2    2023-12-31  19:20:34  이체  내계좌이체   미분류              장보경 -100000  KRW   
3    2023-12-30  18:07:32  지출     식비    배달           이니시스_2   -3000  KRW   
4    2023-12-30  17:40:38  지출     식비   식재료  제이엘  MARKET TOP   -7180  KRW   
...         ...       ...  ..    ...   ...              ...     ...  ...   
1887 2023-01-04  08:38:36  지출     교통    택시         주식회사 티머니   -5600  KRW   
1888 2023-01-04  06:58:57  지출     교통  대중교통           현장발권-S  -18000  KRW   
1889 2023-01-04  04:35:29  지출  문화/여가    구독   NETFLIX_INICIS   -9500  KRW   
1890 2023-01-03  15:21:27  지출  경조/선물    회비              김승연   -9975  KRW   
1891 2023-01-01  08:29:18  수입   금융수입   미분류         입출금통장 이자   18025  KRW   

                          결제수단          메모  
0     KB국민 노리2 체크카드(KB Pay)_교통          #슬  
1             KB Star*t통장-저축예금         NaN  
2                     사이다입출금통장         NaN  
3     KB국민 노리2 체크카드(KB Pay)_교통    치킨배달비 #슬  
4           KB국민 nori 체크카드(RF)          #슬  
...                        ...         ...  
1887                카카오페이 간편결제         NaN  
1888        KB국민 nori 체크카드(RF)  #치앙마이 공항버스  
1889        KB국민 nori 체크카드(RF)         NaN  
1890          KB Star*t통장-저축예금         NaN  
1891                  사이다입출금통장         NaN  

[1892 rows x 10 columns]

---
# 기본 EDA
- 타입: 지출, 이체, 수입 나눠서 데이터 프레임 만들기
- 화폐 열 삭제
- 결제 수단 유니크 확인
- 메모 빈칸은 빈칸으로 두기
- 지출 내역 중복 내역 삭제

In [74]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1892 entries, 0 to 1891
Data columns (total 10 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   날짜      1892 non-null   datetime64[ns]
 1   시간      1892 non-null   object        
 2   타입      1892 non-null   object        
 3   대분류     1892 non-null   object        
 4   소분류     1892 non-null   object        
 5   내용      1892 non-null   object        
 6   금액      1892 non-null   int64         
 7   화폐      1892 non-null   object        
 8   결제수단    1892 non-null   object        
 9   메모      563 non-null    object        
dtypes: datetime64[ns](1), int64(1), object(8)
memory usage: 147.9+ KB


In [14]:
# 타입별 결제수단 확인
df.groupby('타입')['결제수단'].value_counts()

타입  결제수단                    
수입  KB Star*t통장-저축예금            149
    사이다입출금통장                     17
    MY 입출금통장                      8
    OK읏백만통장II                     4
    온라인자립예탁금                      3
    OK짠테크통장                       1
    네이버페이 간편결제                    1
이체  KB Star*t통장-저축예금            177
    사이다입출금통장                    160
    온라인자립예탁금                     40
    카카오페이 머니                     18
    MY 입출금통장                     13
    네이버페이 머니                     13
    KB청년희망적금 5.5 (24.2.24)       12
    청년우대형 주택청약종합저축               10
    KB국민 nori 체크카드(RF)            7
    OK읏백만통장II                     5
    주택청약종합저축                      3
    OK짠테크통장                       1
    예금 4.95 (24.7.14)             1
    KB마이핏통장                       1
    예금 5.5 (24.12.21)             1
    예금 5.1 (24.11.21)             1
    예금 5.0 (24.11.20)             1
지출  KB국민 nori 체크카드(RF)          392
    KB국민 노리2 체크카드(KB Pay)_교통    363
    KB Star*t통장-저축예금            212

In [9]:
#df['화폐'].value_counts()

화폐
KRW    1893
Name: count, dtype: int64

In [76]:
# 화폐 열 삭제
#df.drop('화폐', axis=1, inplace=True)

In [77]:
# 메모 빈칸
df.fillna('', inplace=True)

In [78]:
# 지출 내역 중복 삭제
df = df[~df['내용'].str.contains('티머니', na=False)]

In [79]:
# 지출, 이체, 수입 데이터프레임 분리
지출_df = df[df['타입'] == '지출']
이체_df = df[df['타입'] == '이체']
수입_df = df[df['타입'] == '수입']

지출_df

날짜        시간  타입    대분류    소분류               내용      금액  \
0    2023-12-31  19:21:35  지출     식비     외식        빌라 드 코스테스  -65000   
3    2023-12-30  18:07:32  지출     식비     배달           이니시스_2   -3000   
4    2023-12-30  17:40:38  지출     식비    식재료  제이엘  MARKET TOP   -7180   
5    2023-12-30  14:00:13  지출     식비     외식          굿포유 쌀국수  -17400   
7    2023-12-29  23:19:04  지출     금융  증권/투자              장보경 -100000   
...         ...       ...  ..    ...    ...              ...     ...   
1885 2023-01-04  18:30:37  지출     교통   대중교통           KB카드출금  -42420   
1886 2023-01-04  14:31:30  지출     식비     배달        (주)우아한형제들  -15100   
1888 2023-01-04  06:58:57  지출     교통   대중교통           현장발권-S  -18000   
1889 2023-01-04  04:35:29  지출  문화/여가     구독   NETFLIX_INICIS   -9500   
1890 2023-01-03  15:21:27  지출  경조/선물     회비              김승연   -9975   

                          결제수단          메모  
0     KB국민 노리2 체크카드(KB Pay)_교통          #슬  
3     KB국민 노리2 체크카드(KB Pay)_교통    치킨배달비 #슬  
4           KB국민 nori 체크카드(RF)          #슬  
5           KB국민 nori 체크카드(RF)          #슬  
7                     사이다입출금통장              
...                        ...         ...  
1885          KB Star*t통장-저축예금              
1886        KB국민 nori 체크카드(RF)              
1888        KB국민 nori 체크카드(RF)  #치앙마이 공항버스  
1889        KB국민 nori 체크카드(RF)              
1890          KB Star*t통장-저축예금              

[1214 rows x 9 columns]

---
# 지출_df
- 데이터 프레임 변경 사항
    - '타입' 열 삭제
    - '금액' 데이터 음수 표시 삭제
    - '결제수단' 단순화
    - '세부결제수단' 추가

In [80]:
# '타입'열 삭제
#지출_df.drop('타입', axis=1, inplace=True)

# '금액' 데이터 표시 변경
지출_df['금액'] = 지출_df['금액'].abs()

지출_df

날짜        시간    대분류    소분류               내용      금액  \
0    2023-12-31  19:21:35     식비     외식        빌라 드 코스테스   65000   
3    2023-12-30  18:07:32     식비     배달           이니시스_2    3000   
4    2023-12-30  17:40:38     식비    식재료  제이엘  MARKET TOP    7180   
5    2023-12-30  14:00:13     식비     외식          굿포유 쌀국수   17400   
7    2023-12-29  23:19:04     금융  증권/투자              장보경  100000   
...         ...       ...    ...    ...              ...     ...   
1885 2023-01-04  18:30:37     교통   대중교통           KB카드출금   42420   
1886 2023-01-04  14:31:30     식비     배달        (주)우아한형제들   15100   
1888 2023-01-04  06:58:57     교통   대중교통           현장발권-S   18000   
1889 2023-01-04  04:35:29  문화/여가     구독   NETFLIX_INICIS    9500   
1890 2023-01-03  15:21:27  경조/선물     회비              김승연    9975   

                          결제수단          메모  
0     KB국민 노리2 체크카드(KB Pay)_교통          #슬  
3     KB국민 노리2 체크카드(KB Pay)_교통    치킨배달비 #슬  
4           KB국민 nori 체크카드(RF)          #슬  
5           KB국민 nori 체크카드(RF)          #슬  
7                     사이다입출금통장              
...                        ...         ...  
1885          KB Star*t통장-저축예금              
1886        KB국민 nori 체크카드(RF)              
1888        KB국민 nori 체크카드(RF)  #치앙마이 공항버스  
1889        KB국민 nori 체크카드(RF)              
1890          KB Star*t통장-저축예금              

[1214 rows x 8 columns]

---
---
### 중복 지출 확인
- 체크카드: 티머니 버스.지하철

In [30]:
# # 티머니 버스. 지하철 -> 중복 지출로 잡힘.
# trans = df[df['소분류']=="대중교통"]
# trans.head()

날짜        시간  타입 대분류   소분류          내용     금액              결제수단 메모
6   2023-12-30  07:00:21  지출  교통  대중교통  티머니 버스.지하철  -2900  KB국민 그린체크카드(서울형)   
133 2023-12-05  18:57:16  지출  교통  대중교통      KB카드출금 -34900  KB Star*t통장-저축예금   
289 2023-11-03  18:37:06  지출  교통  대중교통      KB카드출금 -67600  KB Star*t통장-저축예금   
407 2023-10-10  07:15:04  지출  교통  대중교통      KB카드출금 -33065  KB Star*t통장-저축예금   
419 2023-10-06  18:43:54  지출  교통  대중교통      KB카드출금 -31735  KB Star*t통장-저축예금

### 지출 내역 재분류
- 의료/건강 - 운동 을 문화/여가 - 스포츠 로 변경

In [82]:
mask = (지출_df['대분류'] == '의료/건강') & (지출_df['소분류'] == '운동')
지출_df.loc[mask, '대분류'] = '문화/여가'
지출_df.loc[mask, '소분류'] = '스포츠'

---
---
### 결제수단
- 카드
    - KB국민 nori 체크카드(RF)          392   -> 노리 로 변경
    - KB국민 노리2 체크카드(KB Pay)_교통    363   -> 노리 2 로 변경
    - KB국민 그린체크카드(서울형)            160 -> 그린

- 통장
    - KB Star*t통장-저축예금            212 -> 국민스타트
    - 사이다입출금통장                     19 -> 사이다
    - MY 입출금통장                     14 -> 케이뱅크

- 간편결제
    - 카카오페이
        - 카카오페이 간편결제                   42 -> 카페 간편결제
        - 카카오페이 머니                     19 -> 카페 머니

    - 네이버페이
        - 네이버페이 간편결제(머니)               13 -> 네페 머니
        - 네이버페이 간편결제(포인트)              10 -> 네페 포인트
        - 네이버페이 간편결제                    2 -> 네페 간편결제

In [81]:
# 새로운 열 '세부결제수단' 추가 및 값 설정
세부결제수단_매핑 = {
    'KB국민 nori 체크카드(RF)': '노리',
    'KB국민 노리2 체크카드(KB Pay)_교통': '노리 2',
    'KB국민 그린체크카드(서울형)': '그린',
    'KB Star*t통장-저축예금': '국민스타트',
    '사이다입출금통장': '사이다',
    'MY 입출금통장': '케이뱅크',
    '카카오페이 간편결제': '카페 간편결제',
    '카카오페이 머니': '카페 머니',
    '네이버페이 간편결제(머니)': '네페 머니',
    '네이버페이 간편결제(포인트)': '네페 포인트',
    '네이버페이 간편결제': '네페 간편결제'
}

# 세부결제수단 열 추가 및 매핑 적용
지출_df['세부결제수단'] = 지출_df['결제수단'].map(세부결제수단_매핑).fillna('기타')


# 결제수단 열을 '카드', '통장', '간편결제' 중 하나로 설정
지출_df['결제수단'] = 지출_df['결제수단'].map(lambda x: '카드' if '체크카드' in x else ('통장' if '통장' in x else '간편결제'))

# 열 순서 변경 - '메모' 열을 맨 뒤로 이동
열_순서 = ['날짜', '시간', '대분류', '소분류', '내용', '금액', '결제수단', '세부결제수단', '메모']
지출_df = 지출_df[열_순서]

지출_df

날짜        시간    대분류    소분류               내용      금액 결제수단 세부결제수단  \
0    2023-12-31  19:21:35     식비     외식        빌라 드 코스테스   65000   카드   노리 2   
3    2023-12-30  18:07:32     식비     배달           이니시스_2    3000   카드   노리 2   
4    2023-12-30  17:40:38     식비    식재료  제이엘  MARKET TOP    7180   카드     노리   
5    2023-12-30  14:00:13     식비     외식          굿포유 쌀국수   17400   카드     노리   
7    2023-12-29  23:19:04     금융  증권/투자              장보경  100000   통장    사이다   
...         ...       ...    ...    ...              ...     ...  ...    ...   
1885 2023-01-04  18:30:37     교통   대중교통           KB카드출금   42420   통장  국민스타트   
1886 2023-01-04  14:31:30     식비     배달        (주)우아한형제들   15100   카드     노리   
1888 2023-01-04  06:58:57     교통   대중교통           현장발권-S   18000   카드     노리   
1889 2023-01-04  04:35:29  문화/여가     구독   NETFLIX_INICIS    9500   카드     노리   
1890 2023-01-03  15:21:27  경조/선물     회비              김승연    9975   통장  국민스타트   

              메모  
0             #슬  
3       치킨배달비 #슬  
4             #슬  
5             #슬  
7                 
...          ...  
1885              
1886              
1888  #치앙마이 공항버스  
1889              
1890              

[1214 rows x 9 columns]

In [56]:
지출_df['대분류'].value_counts()

대분류
식비       452
카페/간식    308
경조/선물    116
금융        68
의료/건강     43
문화/여가     35
생활        35
교육/학습     34
온라인쇼핑     23
교통        23
뷰티/미용     18
똑순이       17
자동차       14
술/유흥      11
여행/숙박      8
패션/쇼핑      6
주거/통신      4
Name: count, dtype: int64

In [87]:
지출_df.to_excel('2023지출.xlsx', index=False)

## 의료비

In [84]:
# 데이터프레임 만들기
df_med = 지출_df[지출_df['대분류']=='의료/건강']
df_med.head(10)

날짜        시간    대분류    소분류         내용      금액 결제수단 세부결제수단    메모
68  2023-12-19  13:43:22  의료/건강    한의원      치유한의원   13800   카드   노리 2      
180 2023-11-25  11:49:56  의료/건강    한의원      치유한의원   14900   카드     그린      
193 2023-11-22  15:37:22  의료/건강     약국       모아약국   20000   카드   노리 2  흉터연고
196 2023-11-22  11:40:46  의료/건강    한의원      치유한의원   13600   카드   노리 2      
203 2023-11-21  12:36:23  의료/건강   정형외과  월곡순정형외과의원   54300   카드   노리 2      
215 2023-11-20  13:04:34  의료/건강     약국      새국민약국    2200   카드   노리 2      
216 2023-11-20  12:58:50  의료/건강   정형외과  월곡순정형외과의원   51700   카드   노리 2      
393 2023-10-11  17:25:44  의료/건강     치과   서울예담치과의원  288800   카드   노리 2      
401 2023-10-10  16:42:46  의료/건강     약국       부민약국    2700   카드     노리      
402 2023-10-10  16:38:39  의료/건강  이비인후과   김영명이비인후과    5100   카드     노리

In [85]:
df_med_sorted = df_med.sort_values(by='소분류')
df_med_sorted.head()

날짜        시간    대분류   소분류         내용     금액 결제수단 세부결제수단        메모
1664 2023-02-22  20:57:57  의료/건강  보조식품  CJ올리브영_국내  20050   카드     노리  유산균 30*3
1098 2023-06-05  11:55:16  의료/건강    안과       하나안과   7100   카드     노리          
434  2023-10-04  14:47:16  의료/건강    안과       하나안과   7100   카드     노리          
1881 2023-01-05  21:26:26  의료/건강    약국    아이메디칼약국  14000   카드     노리       소화제
989  2023-06-26  11:53:40  의료/건강    약국       한강약국  20500   카드   노리 2       베루말

In [86]:
# 엑셀 파일 저장
df_med.to_excel('병원청구용.xlsx', index=False)

## 슬

- 메모: "#슬" 행만 따로 데이터 프레임 만들기